CSharp/Library/Microsoft.Bot.Connector.Shared/MicrosoftAppCredentials.cs

```diff
@@ -83,7 +83,18 @@ public MicrosoftAppCredentials(IConfiguration configuration, ILogger logger = nu
         public string MicrosoftAppId { get; set; }
         public string MicrosoftAppPassword { get; set; }
 
-        public virtual string OAuthEndpoint { get { return JwtConfig.ToChannelFromBotLoginUrl; } }
+        public virtual string OAuthEndpoint
+        {
+            get
+            {
+                string tenant = null;
+#if NET45
+                // Advanced user only. TODO: add doc link.
+                tenant = SettingsUtils.GetAppSettings("ChannelAuthTenant");
+#endif
+                return string.Format(JwtConfig.ToChannelFromBotLoginUrlTemplate, string.IsNullOrEmpty(tenant) ? "botframework.com" : tenant);
+            }

```

Comment:
*Strongly* consider calling URI.TryCreate on this new string before using it. If the user entered something invalid into the template, and as a result this string is not a valid URI, then their downstream error messages will be impossible to track down. 

Here, it's easy enough to call URI.TryCreate and to throw a very specific error - from just the right place - as to what's going wrong. 

[https://docs.microsoft.com/en-us/dotnet/api/system.uri.trycreate?redirectedfrom=MSDN&view=netframework-4.7.2#System_Uri_TryCreate_System_String_System_UriKind_System_Uri](https://docs.microsoft.com/en-us/dotnet/api/system.uri.trycreate?redirectedfrom=MSDN&view=netframework-4.7.2#System_Uri_TryCreate_System_String_System_UriKind_System_Uri)__

Reply:
good point. done.